File: app/schemas/generic.py
from __future__ import annotations
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime
from app.core.constants import GenericStatus, PregnancyCategory, ControlledSubstanceSchedule


class GenericDrugBase(BaseModel):
    generic_name: str
    chemical_name: Optional[str] = None
    molecular_formula: Optional[str] = None
    cas_number: Optional[str] = None
    therapeutic_class_id: Optional[int] = None
    pharmacology: Optional[str] = None
    indications: Optional[str] = None
    contraindications: Optional[str] = None
    side_effects: Optional[str] = None
    interactions: Optional[str] = None
    pregnancy_category: Optional[PregnancyCategory] = None
    controlled_substance_schedule: ControlledSubstanceSchedule = ControlledSubstanceSchedule.NONE
    status: GenericStatus = GenericStatus.ACTIVE


class GenericDrugCreate(GenericDrugBase):
    pass


class GenericDrugUpdate(BaseModel):
    generic_name: Optional[str] = None
    chemical_name: Optional[str] = None
    molecular_formula: Optional[str] = None
    cas_number: Optional[str] = None
    therapeutic_class_id: Optional[int] = None
    pharmacology: Optional[str] = None
    indications: Optional[str] = None
    contraindications: Optional[str] = None
    side_effects: Optional[str] = None
    interactions: Optional[str] = None
    pregnancy_category: Optional[PregnancyCategory] = None
    controlled_substance_schedule: Optional[ControlledSubstanceSchedule] = None
    status: Optional[GenericStatus] = None


class GenericDrugInDB(GenericDrugBase):
    generic_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class GenericDrugResponse(GenericDrugInDB):
    therapeutic_class_name: Optional[str] = None


class GenericDrugWithDetails(GenericDrugResponse):
    therapeutic_class: Optional["TherapeuticClassResponse"] = None
    brand_count: int = 0

File: frontend/src/types/generic.ts
export interface GenericDrug {
  generic_id: number;
  generic_name: string;
  chemical_name?: string;
  molecular_formula?: string;
  cas_number?: string;
  therapeutic_class_id?: number;
  therapeutic_class_name?: string;
  pharmacology?: string;
  indications?: string;
  contraindications?: string;
  side_effects?: string;
  interactions?: string;
  pregnancy_category?: string;
  controlled_substance_schedule?: string;
  status?: string;
  created_at?: string;
  updated_at?: string;
  // Legacy fields for compatibility
  id?: number;
  scientific_name?: string;
  arabic_name?: string;
  description?: string;
}

export interface GenericDrugWithDetails extends GenericDrug {
  therapeutic_class?: {
    id: number;
    name: string;
  };
  brands?: BrandSummary[];
  brand_count?: number;
}

export interface BrandSummary {
  id: number;
  name: string;
  manufacturer_name?: string;
}

export interface GenericDrugCreate {
  generic_name: string;
  chemical_name?: string;
  molecular_formula?: string;
  cas_number?: string;
  therapeutic_class_id?: number;
  pharmacology?: string;
  indications?: string;
  contraindications?: string;
  side_effects?: string;
  interactions?: string;
  pregnancy_category?: string;
  controlled_substance_schedule?: string;
  status?: string;
}

export interface GenericDrugUpdate {
  generic_name?: string;
  chemical_name?: string;
  molecular_formula?: string;
  cas_number?: string;
  therapeutic_class_id?: number;
  pharmacology?: string;
  indications?: string;
  contraindications?: string;
  side_effects?: string;
  interactions?: string;
  pregnancy_category?: string;
  controlled_substance_schedule?: string;
  status?: string;
}
